python/lsst/ts/hexrotcomm/command_telemetry_client.py

```diff
@@ -65,46 +75,53 @@ class CommandTelemetryClient:
         await client.stop()
     """
 
-    telemetry_interval = 0.1
-    """Interval between telemetry messages (sec)."""
-
-    connect_retry_interval = 0.1
-    """Interval between connection retries (sec)."""
-
     def __init__(
         self,
         *,
         log,
-        config,
-        telemetry,
+        ConfigClass,
+        TelemetryClass,
+        host,
         command_port,
         telemetry_port,
-        host=tcpip.LOCAL_HOST,
+        connect_callback,
+        config_callback,
+        telemetry_callback,
+        connect_timeout=10,
     ):
         self.log = log.getChild("BaseMockController")
-        self.config = config
-        self.telemetry = telemetry
+        self.header = structs.Header()
+        self.config = ConfigClass()
+        self.telemetry = TelemetryClass()
         self.command_port = command_port
         self.telemetry_port = telemetry_port
         self.host = host
-        self.headers = dict()
-        # A dictionary of frame ID: header for telemetry and config data
-        # Keeping separate headers for telemetry and config allows
-        # updating just the relevant fields, rather than creating a new
-        # header insteance for each telemetry and config message.
-        for frame_id in (config.FRAME_ID, telemetry.FRAME_ID):
-            header = structs.Header()
-            header.frame_id = frame_id
-            self.headers[frame_id] = header
+        self.connect_callback = connect_callback
+        self.config_callback = config_callback
+        self.telemetry_callback = telemetry_callback
+        self.connect_timeout = connect_timeout
         self.command_reader = None
         self.command_writer = None  # not written
         self.telemetry_reader = None  # not read
         self.telemetry_writer = None
-        self.command_loop_task = asyncio.Future()
-        self.command_loop_task.set_result(None)
-        self.telemetry_loop_task = asyncio.Future()
-        self.telemetry_loop_task.set_result(None)
         self.connect_task = asyncio.create_task(self.connect())
+        # The state of (command_connected, telemetry_connected)
+        # the last time connect_callback was called.
+        self.was_connected = (False, False)
+        self.should_be_connected = False
+
+        # Dict of command_code: number of times this command has been sent;
+        # note that the count will wrap around at some point determined
+        # by the data type of Command.counter.
+        self._command_counts = dict()
+
+        # Task the user can set to an asyncio.Future()
+        # in order to detect when the next telemetry is read.
+        # Only intended for unit tests.
+        self._telemetry_task = asyncio.Future()
+
+        self.read_telemetry_and_config_task = utils.make_done_future()
+        self.monitor_command_reader_task = utils.make_done_future()
```

Comment:
Did you add the doc string for these two attributes in `Parameters`? Actually, I felt there are too many attributes in this class right now...

Reply:
Yes. I agree. I made some private and documented the rest in a new Attributes section.